### ISSUE TITLE
Turret model

### ISSUE THREAD

**Issue by Pyr0xene** [mwpenny/portal64-still-alive#97]
Textured and rigged turret model with LODs

![image](https://github.com/user-attachments/assets/1465c86d-e885-4c43-a658-65d5ff724aac)
![image](https://github.com/user-attachments/assets/9d64e973-7b5f-4d47-a4aa-f2a414cf85b6)

(Apologies if I'm formatting this wrong in any way)
So about a year ago I was commissioned by a certain someone who shall remain nameless :) to model a Portal turret according to N64 specifications. Today I saw that some progress had already been made on a turret model in this project, which is looking good. But in case you wanted to save some time, I'd gladly contribute this ready-made one.

The lod0 model has 369 tris, lod1 and lod2 have 184 and 98 respectively. Textures are two 64x32 ci4 tiles so they'll fit in tmem together. Rig has 1 bone for the body and 1 bone each for the guns.

**Comment by mwpenny** [comment]
Very much appreciated. You've done a great job here! The different levels of detail are a nice touch.

Before this can be committed to the repository, textures will need to be removed from the model file and instead programmatically generated from the original Portal game files. I can do this cleanup.

What is the easiest way to get in touch to send the model file? For example Discord or Email. 

Thank you!

**Comment by Pyr0xene** [comment]
Just added my discord to my bio, we can use that!

**Comment by mwpenny** [comment]
For others who may be following this - I have the models and will work on integrating them into the game. It looks like some adjustments may be needed, and I will measure performance.

Thanks @Pyr0xene! I'll keep this issue open to track progress.

**Comment by hackgrid** [comment]
Very nice, I am happy that these efforts were not wasted! Looking forward
to see the turret ingame very much :)

Matt Penny ***@***.***> schrieb am Do., 19. Dez. 2024, 10:56:

**Comment by mwpenny** [comment]
Had some time today to work on this and made some good progress (placing turrets in levels using Blender, loading and rendering in game).

It will take a bit more work before it will look right:
* These are the first models in the game using color indexed textures and the build isn't set up to include the palette. Skeletool supports reading external palettes though so will just take some tinkering. I've been hard-coding it manually while I set things up.
* The model scale/dimensions will need to be adjusted. As-is, they are quite a bit larger than the player. Portal 64 uses 1/64 scale of the original game.
* The turret is in the style of Portal 2. Switching to Portal 1 style could be done later, although doing so could solve some other problems (for example, could maybe fit everything in 1 texture and/or reduce model complexity)

I'll have some more time tomorrow and after Christmas.

**Comment by hackgrid** [comment]
Nice! :) Do you plan to push your current state or will you wait until you are finished / pleased enough with your work? :)
I did not notice the differences between Portal 1 and Portal 2 turrents until now :D

**Comment by mwpenny** [comment]
I'm going to get it to a bit better of a state first. I'd like to at least fix the scale/dimensions and get it rendering with textures without people having to hack around it like I did.

The main differences are the inside of the turret. Portal 1 has more solid colors and more of the internals are covered. Portal 2 adds more details. That's good news for us though since making it more faithful to the original also makes it simpler :)

**Comment by mwpenny** [comment]
I should have expected that something new like this would uncover bugs :)

* Skeletool didn't properly handle exporting 4-bit color indexed textures (both in terms of the data written out, and how it loaded it into the N64's texture memory)
* Skeletool didn't support multiple palettes in tmem at the same time
* The game didn't support textures on dynamically loaded models. I ended up not using this (so multiple turrets can all be drawn while keeping the same textures in memory), but fixed it since it's useful for the future

With everything fixed, the lod0 model is in the game. I've placed one at the start of chamber 11. I adjusted the dimensions/scale and they still don't seem quite right. But it's good enough for now.

Next steps are to test performance and go from there, and adjust to the Portal 1 style. Soon I'll have a better flashcart to more easily/quickly debug and test things on console. Hopefully we don't have to compromise on detail here - although at original resolution much of it is hidden.

**Comment by hackgrid** [comment]
Cool, great work and happy new year! :)

I own almost every flashcard (64drive, Everdrive, summerCart) and I wanted
to do a tutorial for each how to use it and debug e.g. with ares gdb server
in the future :) (though I have to figure that for me first :D)

Matt Penny ***@***.***> schrieb am Mi., 1. Jan. 2025, 00:31:

**Comment by mwpenny** [comment]
Happy new year! It being on the ground is not intended haha. Is this on Ares?

**Comment by hackgrid** [comment]
Yes, check my other new issue for details :)

Matt Penny ***@***.***> schrieb am Mi., 1. Jan. 2025, 16:42:

**Comment by hackgrid** [comment]
Cool, awesome work!
Maybe there is still some potential for optimizations with portal64?

I know all people keep mention Kaze all the time, but after that all eyecandy he added to SM64, he managed to get close to 60FPS now.

Less triangles will work of course, but maybe there are other ways of optimization still left?
Maybe you could see how it plays out first, I remember James did not like premature optimization and only optimized stuff when it was needed - so that is why I think there should be some fat too trim still?

I do not know nothing about 3D programming of course, so unfortunately I can not help at all and may be way off with that opinion, sorry!

**Comment by mwpenny** [comment]
There's definitely room to improve the engine further. What I'm trying to avoid here is making this just barely doable. Building in a bit of headroom (where possible) each time new content is added leaves space for other complexities down the line.

Kaze's Mario 64 hack is impressive, though he has been optimizing it for years at this point. Some of the performance is also due to assembly-level hacking and custom microcode (libdragon + something like tiny3d may help here). So it will take time to achieve a similar level.

With that said, I've already found and implemented some performance improvements and I know there will be more. You're right that they were the types of things that just weren't important early on. I agree we shouldn't over/prematurely optimize, but trimming a bit of fat along the way avoids unnecessary accumulation / gains more stomach space / etc. :)

Here it feels worth at least looking into. Right now the lod0 turret has almost as many triangles as the player model. Given the performance difference and how lod1 looks at N64 resolution, I'm confident a lower poly turret can look very good. Regardless, I will experiment before making changes. I don't want to spend forever on this either - turret AI will be quite fun.

**Comment by Pyr0xene** [comment]
Amazing work getting it in-game! I really like how lowpoly models look at low res and composite output, I actually prefer it haha.

Sorry I've been absent, but I can help out some more now if you'd like. I can adjust the model to resemble the Portal 1 turret (I never even realized they were different) and I'm definitely seeing some easy gains in polycount and especially texture resolution right away just from the different design, even before additional optimization.

I've also learned a lot about N64 materials since making this over a year ago, so if it's possible to use a custom color combiner setup (as well as setting parameters such as prim/env color etc) in your build, I could use that as well.

**Comment by mwpenny** [comment]
I agree, It's a very cool aesthetic! Your models look right at home.

> Sorry I've been absent, but I can help out some more now if you'd like.

No problem on being away, there's no pressure on amount to contribute. The changes you mentioned sound great! If you did that, I could shift my current focus to turret gameplay and in parallel we could get chamber 16 fleshed out more quickly.

While you're looking at this - Portal 64 uses 1/64 scale dimensions of the original game. I roughly adjusted the scale of the current models but haven't gone over it in detail and so I think it's still not quite accurate.

Also, if possible, it's easier to integrate this into the build if original textures are used and then edited as necessary. Then the repo would contain textureless models + commands to turn original-game textures into N64-ready ones, which is how most other textures work (`.ims` files in `assets/materials/` define edits, and resulting textures are output to `portal_pak_modified/materials/` when building the game). I think this is more doable with the simpler design of the Portal 1 turrets but could be wrong. If it gives you trouble I could help out.  Or if it doesn't seem feasible then just make sure (like before) the new textures are fully original and not sourced from the originals in some way.

> I've also learned a lot about N64 materials since making this over a year ago, so if it's possible to use a custom color combiner setup (as well as setting parameters such as prim/env color etc) in your build, I could use that as well.

I've recently documented the asset pipeline. Materials are [here](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/assets/materials.md) and models are [here](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/assets/models.md). I've not documented the material format itself since that will eventually change.

Materials are defined in YAML files containing objects which represent libultra display list commands. Since these are just wrappers for display list commands, it's straightforward to set the primitive/environment color and use a custom color combiner setup. For example: https://github.com/mwpenny/portal64-still-alive/blob/88e35bbe91154aa322991c79ffde078f45efc6cc/assets/materials/static.skm.yaml#L104-L115

For models, just make sure the material name in Blender matches the name in the YAML file.

Also, as an optimization to avoid texture switching it's possible to define a "default" material for a model. When generating a model's display list, the default material is assumed to have been already applied and so any redundant commands will be excluded from the generated display list.

For example, the turret models use two materials: `turret1` and `turret2`. But a default material containing everything is defined [here](https://github.com/mwpenny/portal64-still-alive/blob/88e35bbe91154aa322991c79ffde078f45efc6cc/assets/materials/static.skm.yaml#L1348-L1370) in `static.skm.yaml`, so the engine is able to [tag all turrets](https://github.com/mwpenny/portal64-still-alive/blob/88e35bbe91154aa322991c79ffde078f45efc6cc/src/scene/turret.c#L33) with it. This allows them to be rendered together and the default material is just loaded once beforehand to set up the necessary RCP state. The actual materials referenced by the models are defined [here](https://github.com/mwpenny/portal64-still-alive/blob/88e35bbe91154aa322991c79ffde078f45efc6cc/assets/materials/objects.skm.yaml#L253-L288) in `objects.skm.yaml`. They each define what the state of the RCP needs to be, but due to the default material, the only thing that really changes is the current tile being used (`gSPTexture()`).

**Comment by hackgrid** [comment]
Now turrets seem kinda finished and they are awesome! :)
Thanks for the great work @mwpenny !

**Comment by mwpenny** [comment]
That's okay, life happens! Sounds great - whenever you have time.

Just wanted to avoid duplicating effort (and your art skills are better than mine :p)

**Comment by Pyr0xene** [comment]
The march of time still keeps catching me off guard, my god. 

Here's the portal 1 version of the turret. I'm actually way happier with it - the texture space saved on the internal detail allowed some weathering and edge detail on the outer shell itself, looks way closer to the PC version. And it's only using one CI4 texture this time. The texture is also laid out horizontally at 128x32 which I've read is more efficient on the N64, and it was convenient anyway.

The material is 2cycle (according to Kaze this is ok and sometimes even faster??). 
I actually stole this from the Playstation, it's the same formula PSX uses to blend vertex color with texture color. It allows for some nice overbright without washing out detail (as long as care is taken not to let it overflow of course): 
```
[TEX0, 0, SHADE, 0],
[1, 0, COMBINED, COMBINED]

```
TEX0 should use mirroring on both axes.

![Image](https://github.com/user-attachments/assets/c92cc0cf-f7fd-47be-a74a-89dd13fb13bf)

![Image](https://github.com/user-attachments/assets/4102bbbf-a855-486b-8bd1-293d22a705cd)

LOD0 is down to 350 tris from 369, LOD1 and LOD2 are now 196 and 114, which is a handful more than they were but they also match LOD0 much better now, so I'm hoping they should be able to be brought in a bit closer to the camera without the change being as jarring.

I did attempt to see if we could use the original Portal textures somehow with this version, as suggested earlier, but there's just no way haha. Not only is the UV space much too different with deformation etc that's needed for optimization to get anywhere near an acceptable resolution, there's also the fact that a lot of fine detail in the original models is 3D geometry and not represented in the textures, and as such need to be drawn or baked in, like the thin metal pipes/wiring on the inside (which also need to be pixel-aligned and that's not always possible without doing it by hand and taking some artistic liberties) and the guns. Lastly the color reduction for CI4 really needs manual pixel-level work for textures like these, it just won't look good otherwise. So this texture is drawn by hand from scratch while eyeballing the original model, which I hope will be fine to include in the project?

There's probably a lot of other, simpler props where these issues can be avoided though.

Anyways, I'll contact @mwpenny on discord to send the files over!

**Comment by hackgrid** [comment]
very cool, thanks for your hard work! :)

**Comment by mwpenny** [comment]
> The march of time still keeps catching me off guard, my god.

Time seriously flies. No worries though, I appreciate your work on this. The models look great! The texture reduction and details are a nice touch.

> The material is 2cycle (according to Kaze this is ok and sometimes even faster??).

I assume you're referring to [this video](https://youtu.be/YJJC6FJKu58?si=orx9cPEDccLlkKos&t=2018). Kaze claims it is "~5% cheaper under the right conditions but also ~30% more expensive under bad conditions". The video shows "depth test failed" above the "bad condition" results and so it seems like one of the contributing factors is occlusion or overdraw. But he also doesn't go into too much detail and admits that it's all a bit mysterious anyway.

Worst case, we can use 1 cycle mode instead (no highlight). But the highlights achieved with your color combiner setup look nice. I'll measure performance and see what kind of impact this has.

> LOD0 is down to 350 tris from 369, LOD1 and LOD2 are now 196 and 114, which is a handful more than they were but they also match LOD0 much better now, so I'm hoping they should be able to be brought in a bit closer to the camera without the change being as jarring.

So far, the Portal 2 style LOD0 model has been performing reasonably well in "standard" playthroughs of the level (not trying to tank the framerate). I'm optimistic for these new models as well. I may not need to add LOD switching until extra detail is added to chamber 16 (more props + behind-the-scenes area), or even chamber 18/19 :) In any case, the tri reduction and model consistency will definitely help.

> I did attempt to see if we could use the original Portal textures somehow with this version, as suggested earlier, but there's just no way haha. [...] So this texture is drawn by hand from scratch while eyeballing the original model, which I hope will be fine to include in the project?

No problem. Thanks for trying anyway and explaining the drawbacks. Since your texture is original and not ripped/derived from the game files, this is fine.

Thanks again and I'll look at getting this in-game!

**Comment by mwpenny** [comment]
The new (old) turrets are in and they look great. While comparing these to the PC version I noticed some artistic touches I never have before. Excellent attention to detail.

![Image](https://github.com/user-attachments/assets/01c00dd8-b88b-47c0-835b-f1d881cbf610)

I redid the stress tests I described above. Performance is comparable to the old models in most cases, and slightly better at times (marked with *). Again, this isn't perfect but gives a rough idea.

|                | 1 turret   | 3 turrets  | 6 turrets  | 9 turrets  | 12 turrets |
| -------------- | ---------- | ---------- | ---------- | ---------- | ---------- |
| lod0           | 60         | 30         | 30         | 30         | 30         |
| lod0 + portals | 30         | 30         | 20         | **20***    | **15***    |
| lod1           | 60         | 60         | 30         | 30         | 30         |
| lod1 + portals | 30         | 30         | 30         | 20         | 20         |
| lod2           | 60         | 60         | 30         | 30         | 30         |
| lod2 + portals | 30         | 30         | 30         | **30***    | 20         |

I will implement LOD switching eventually, which will help in busy scenes, but not for v0.17.0. 

With that, I think this issue can be closed. Great work!
